# Exploração e tratamento de dados
- Carregamento de base de dados
- Análise da consistencia dos dados
- Tratamento de dados

In [1]:
# bibliotecas
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

## Exploração

In [2]:
# carregando dados
base_credit = pd.read_csv('/data/credit_data.csv')
base_credit

clientid        income        age         loan  default
0            1  66155.925095  59.017015  8106.532131        0
1            2  34415.153966  48.117153  6564.745018        0
2            3  57317.170063  63.108049  8020.953296        0
3            4  42709.534201  45.751972  6103.642260        0
4            5  66952.688845  18.584336  8770.099235        1
...        ...           ...        ...          ...      ...
1995      1996  59221.044874  48.518179  1926.729397        0
1996      1997  69516.127573  23.162104  3503.176156        0
1997      1998  44311.449262  28.017167  5522.786693        1
1998      1999  43756.056605  63.971796  1622.722598        0
1999      2000  69436.579552  56.152617  7378.833599        0

[2000 rows x 5 columns]

In [ ]:
# primeiras linhas
base_credit.head()

clientid        income        age         loan  default
0         1  66155.925095  59.017015  8106.532131        0
1         2  34415.153966  48.117153  6564.745018        0
2         3  57317.170063  63.108049  8020.953296        0
3         4  42709.534201  45.751972  6103.642260        0
4         5  66952.688845  18.584336  8770.099235        1

In [ ]:
# ultimas linhas
base_credit.tail()

clientid        income        age         loan  default
1995      1996  59221.044874  48.518179  1926.729397        0
1996      1997  69516.127573  23.162104  3503.176156        0
1997      1998  44311.449262  28.017167  5522.786693        1
1998      1999  43756.056605  63.971796  1622.722598        0
1999      2000  69436.579552  56.152617  7378.833599        0

In [ ]:
# dimensoes
base_credit.shape

(2000, 5)

In [ ]:
# atributos
base_credit.columns

Index(['clientid', 'income', 'age', 'loan', 'default'], dtype='object')

In [ ]:
# informacoes do dataset
base_credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   clientid  2000 non-null   int64  
 1   income    2000 non-null   float64
 2   age       1997 non-null   float64
 3   loan      2000 non-null   float64
 4   default   2000 non-null   int64  
dtypes: float64(3), int64(2)
memory usage: 78.3 KB


*todos atributos são numéricos e podemos confirmasr que o atributo 'age' apresenta 3 valores faltantes.*

In [ ]:
# descritiva
base_credit.describe()

clientid        income          age          loan      default
count  2000.000000   2000.000000  1997.000000   2000.000000  2000.000000
mean   1000.500000  45331.600018    40.807559   4444.369695     0.141500
std     577.494589  14326.327119    13.624469   3045.410024     0.348624
min       1.000000  20014.489470   -52.423280      1.377630     0.000000
25%     500.750000  32796.459717    28.990415   1939.708847     0.000000
50%    1000.500000  45789.117313    41.317159   3974.719419     0.000000
75%    1500.250000  57791.281668    52.587040   6432.410625     0.000000
max    2000.000000  69995.685578    63.971796  13766.051239     1.000000

clientid: com ID de 1 a 2000, isso indica presença de id único.

income: com média em 45.332, a renda varia entre 20.014 e 69.996, 50% dos clientes tem renda inferior a 45.789 já 75% dos clientes tem renda inferior 57.791, uma distribuição não uniforme onde a maioria dos clientes concentrados em rendas mais baixas.

age: com idade média em 40 anos, devido a valores fora do domínio as análises de amplitude são desconsideradas antes de tratamento por apresentar valores negativos.
podemos observar que 50% dos clientes possuem idade inferior a 41 anos.

loan: com valores médios em 4.444,37 e um desvio padrão de 3045.41 e valores que vão de 1.38 a 13766.05 percebe-se dispersão nos dados, metade dos clientes tem emprestimos inferiores a 3974.72.

default: a media de inadimplencia é de 14.15%


*obs: atributo 'age' apresenta valores fora do domínio com idades negativas.*

In [ ]:
# valores faltantes
base_credit.isnull().sum()

clientid    0
income      0
age         3
loan        0
default     0
dtype: int64

In [ ]:
# clientes com idades negativas
base_credit[base_credit['age'] < 0]

clientid        income        age         loan  default
15        16  50501.726689 -28.218361  3977.287432        0
21        22  32197.620701 -52.423280  4244.057136        0
26        27  63287.038908 -36.496976  9595.286289        0

In [ ]:
# clientes com valores faltantes
base_credit[base_credit['age'].isnull()]

clientid        income  age         loan  default
28        29  59417.805406  NaN  2082.625938        0
30        31  48528.852796  NaN  6155.784670        0
31        32  23526.302555  NaN  2862.010139        0

Confirmados 6 registros para tratamento do atributo age.

In [ ]:
# confirmando dados unicos do atributo 'default'
# base_credit['default'].unique()
# base_credit['default'].value_counts()
# np.unique(base_credit['default'])
np.unique(base_credit['default'], return_counts=True)

(array([0, 1]), array([1717,  283]))

confirma-se que inadimplentes são a minoria e os atributos representam uma classe binária [0,1]

In [ ]:
# cliente com maior renda
# base_credit[base_credit['income'] == base_credit['income'].max()]
# base_credit.loc[base_credit['income'].idxmax()]
base_credit.query('income == income.max()')

clientid        income        age         loan  default
422       423  69995.685578  52.719673  2084.370861        0

In [ ]:
# cliente com menor emprestimo
# base_credit[base_credit['loan'] == base_credit['loan'].min()]
# base_credit.loc[base_credit['loan'].idxmin()]
base_credit.query('loan == loan.min()')

clientid        income        age     loan  default
865       866  28072.604355  54.142548  1.37763        0

## Tratamento

**inconsistentes**

In [5]:
# media sem dados inconsistentes em 'age'
base_credit.loc[base_credit['age'] > 0, 'age'].mean()

40.92770044906149

In [10]:
# indices para confirmacao
base_credit[base_credit['age'] < 0].index

Index([15, 21, 26], dtype='int64')

In [11]:
# atribuindo media aos inconsistentes
base_credit.loc[base_credit['age'] < 0, 'age'] = 40.92

In [14]:
# confirmando tratamento
base_credit.iloc[[15,21,26]]

clientid        income    age         loan  default
15        16  50501.726689  40.92  3977.287432        0
21        22  32197.620701  40.92  4244.057136        0
26        27  63287.038908  40.92  9595.286289        0

In [19]:
# check
(base_credit['age'] < 0).any()

False

**nulos**

In [20]:
# valores nulos
base_credit[base_credit['age'].isnull()]

clientid        income  age         loan  default
28        29  59417.805406  NaN  2082.625938        0
30        31  48528.852796  NaN  6155.784670        0
31        32  23526.302555  NaN  2862.010139        0

In [21]:
# indices
base_credit[base_credit['age'].isnull()].index

Index([28, 30, 31], dtype='int64')

In [25]:
# atribuindo a media
# base_credit.loc[base_credit['age'].isnull(), 'age'] = base_credit['age'].mean()
base_credit['age'].fillna(base_credit['age'].mean(), inplace=True)

<ipython-input-25-2d7b7ee5da3f>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  base_credit['age'].fillna(base_credit['age'].mean(), inplace=True)


In [26]:
# conferindo nulos afetados
base_credit.iloc[[28,30,31]]

clientid        income        age         loan  default
28        29  59417.805406  40.927689  2082.625938        0
30        31  48528.852796  40.927689  6155.784670        0
31        32  23526.302555  40.927689  2862.010139        0

In [27]:
# check
base_credit['age'].isnull().any()

False